File: LeetCode/983.py
class Solution:            
    def mincostTickets(self, days: List[int], costs: List[int]) -> int:
        n = len(days)
        memo = dict()
        def DP(i):
            nonlocal days, costs, n, memo
            if i >= n: return 0
            val = memo.get(i)
            if val: return val
            plus30 = plus7 = 1
            current_day = days[i]
            while i + plus7 < n and days[i+plus7] < current_day + 7:
                plus7 += 1
            while i + plus30 < n and days[i+plus30] < current_day + 30:
                plus30 += 1
            res = memo[i] = min(costs[0] + DP(i+1), min(costs[1] + DP(i+plus7), costs[2] + DP(i+plus30)))
            return res
        return DP(0)

File: LeetCode/983.cc
const int INF = 1e9;
class Solution {
    vector<int> memo;
    
    int dp(int i, const vector<int>& days, const vector<int>& costs) {
        if(i == days.size()) return 0;
        if(int& ans = memo[i]; ans != INF) return ans;
        else {
            int next_7 = i+1;
            for(; next_7 < days.size(); ++next_7) if(days[next_7] >= days[i]+7) break;
            int next_30 = i+1;
            for(; next_30 < days.size(); ++next_30) if(days[next_30] >= days[i]+30) break;
            return ans = min({costs[0] + dp(i+1, days, costs), 
                              costs[1] + dp(next_7, days, costs),
                              costs[2] + dp(next_30, days, costs)});
        }
    }
    
public:
    int mincostTickets(vector<int>& days, vector<int>& costs) {
        memo.assign(366, INF);
        return dp(0, days, costs);
    }
};